File: lua/test/get_postal_codes_by_city_entity_test.lua
-- GetPostalCodesByCity entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("zippopotamus-zip-code_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("GetPostalCodesByCityEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:GetPostalCodesByCity(nil)
    assert.is_not_nil(ent)
  end)

  -- Feature #4: the entity stream(action, ...) method runs the op pipeline and
  -- returns an iterator over result items. With the streaming feature active it
  -- yields the feature's incremental output; otherwise it falls back to the
  -- materialised list so stream always yields.
  it("should stream", function()
    local seed = {
      entity = {
        ["get_postal_codes_by_city"] = {
          s1 = { id = "s1" },
          s2 = { id = "s2" },
          s3 = { id = "s3" },
        },
      },
    }

    -- Fallback: streaming inactive -> yields the materialised list items.
    local base = sdk.test(seed, nil)
    local seen = {}
    for item in base:GetPostalCodesByCity(nil):stream("list", nil, nil) do
      table.insert(seen, item)
    end
    assert.are.equal(3, #seen)

    -- Inbound: streaming active -> yields each item from the feature.
    local config = require("config")()
    if type(config.feature) == "table" and config.feature.streaming ~= nil then
      local streamsdk = sdk.test(seed, { feature = { streaming = { active = true } } })
      local got = {}
      for item in streamsdk:GetPostalCodesByCity(nil):stream("list", nil, nil) do
        if vs.islist(item) then
          for _, sub in ipairs(item) do
            table.insert(got, sub)
          end
        else
          table.insert(got, item)
        end
      end
      assert.are.equal(3, #got)
    end
  end)

  it("should run basic flow", function()
    local setup = get_postal_codes_by_city_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"list"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "get_postal_codes_by_city." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set ZIPPOPOTAMUS_ZIP_CODE_TEST_GET_POSTAL_CODES_BY_CITY_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- Bootstrap entity data from existing test data.
    local get_postal_codes_by_city_ref01_data_raw = vs.items(helpers.to_map(
      vs.getpath(setup.data, "existing.get_postal_codes_by_city")))
    local get_postal_codes_by_city_ref01_data = nil
    if #get_postal_codes_by_city_ref01_data_raw > 0 then
      get_postal_codes_by_city_ref01_data = helpers.to_map(get_postal_codes_by_city_ref01_data_raw[1][2])
    end

    -- LIST
    local get_postal_codes_by_city_ref01_ent = client:GetPostalCodesByCity(nil)
    local get_postal_codes_by_city_ref01_match = {
      ["city"] = setup.idmap["city01"],
      ["country"] = setup.idmap["country01"],
      ["state"] = setup.idmap["state01"],
    }

    local get_postal_codes_by_city_ref01_list_result, err = get_postal_codes_by_city_ref01_ent:list(get_postal_codes_by_city_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(get_postal_codes_by_city_ref01_list_result)

  end)
end)

function get_postal_codes_by_city_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/get_postal_codes_by_city/GetPostalCodesByCityTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read get_postal_codes_by_city test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "get_postal_codes_by_city01", "get_postal_codes_by_city02", "get_postal_codes_by_city03", "city01", "country01", "state01" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("ZIPPOPOTAMUS_ZIP_CODE_TEST_GET_POSTAL_CODES_BY_CITY_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["ZIPPOPOTAMUS_ZIP_CODE_TEST_GET_POSTAL_CODES_BY_CITY_ENTID"] = idmap,
    ["ZIPPOPOTAMUS_ZIP_CODE_TEST_LIVE"] = "FALSE",
    ["ZIPPOPOTAMUS_ZIP_CODE_TEST_EXPLAIN"] = "FALSE",
  })

  local idmap_resolved = helpers.to_map(
    env["ZIPPOPOTAMUS_ZIP_CODE_TEST_GET_POSTAL_CODES_BY_CITY_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["ZIPPOPOTAMUS_ZIP_CODE_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["ZIPPOPOTAMUS_ZIP_CODE_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["ZIPPOPOTAMUS_ZIP_CODE_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end

File: rb/test/get_postal_codes_by_city_entity_test.rb
# GetPostalCodesByCity entity test

require "minitest/autorun"
require "json"
require_relative "../ZippopotamusZipCode_sdk"
require_relative "runner"

class GetPostalCodesByCityEntityTest < Minitest::Test
  def test_create_instance
    testsdk = ZippopotamusZipCodeSDK.test(nil, nil)
    ent = testsdk.GetPostalCodesByCity(nil)
    assert !ent.nil?
  end

  # Feature #4: the entity stream(action, ...) method runs the op pipeline and
  # returns an Enumerator over result items. With the streaming feature active
  # it yields the feature's incremental output; otherwise it falls back to the
  # materialised list so stream always yields.
  def test_stream
    seed = {
      "entity" => {
        "get_postal_codes_by_city" => {
          "s1" => { "id" => "s1" },
          "s2" => { "id" => "s2" },
          "s3" => { "id" => "s3" },
        },
      },
    }

    # Fallback: streaming inactive -> yields the materialised list items.
    base = ZippopotamusZipCodeSDK.test(seed, nil)
    seen = base.GetPostalCodesByCity(nil).stream("list", nil, nil).to_a
    assert_equal 3, seen.length

    # Inbound: streaming active -> yields each item from the feature.
    cfg = ZippopotamusZipCodeConfig.shared_config
    if cfg["feature"].is_a?(Hash) && cfg["feature"].key?("streaming")
      sdk = ZippopotamusZipCodeSDK.test(seed, { "feature" => { "streaming" => { "active" => true } } })
      got = []
      sdk.GetPostalCodesByCity(nil).stream("list", nil, nil).each do |item|
        if item.is_a?(Array)
          got.concat(item)
        else
          got << item
        end
      end
      assert_equal 3, got.length
    end
  end

  def test_basic_flow
    setup = get_postal_codes_by_city_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["list"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "get_postal_codes_by_city." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set ZIPPOPOTAMUS_ZIP_CODE_TEST_GET_POSTAL_CODES_BY_CITY_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    get_postal_codes_by_city_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.get_postal_codes_by_city")))
    get_postal_codes_by_city_ref01_data = nil
    if get_postal_codes_by_city_ref01_data_raw.length > 0
      get_postal_codes_by_city_ref01_data = Helpers.to_map(get_postal_codes_by_city_ref01_data_raw[0][1])
    end

    # LIST
    get_postal_codes_by_city_ref01_ent = client.GetPostalCodesByCity(nil)
    get_postal_codes_by_city_ref01_match = {
      "city" => setup[:idmap]["city01"],
      "country" => setup[:idmap]["country01"],
      "state" => setup[:idmap]["state01"],
    }

    get_postal_codes_by_city_ref01_list_result = get_postal_codes_by_city_ref01_ent.list(get_postal_codes_by_city_ref01_match, nil)
    assert get_postal_codes_by_city_ref01_list_result.is_a?(Array)

  end
end

def get_postal_codes_by_city_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "get_postal_codes_by_city", "GetPostalCodesByCityTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = ZippopotamusZipCodeSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["get_postal_codes_by_city01", "get_postal_codes_by_city02", "get_postal_codes_by_city03", "city01", "country01", "state01"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["ZIPPOPOTAMUS_ZIP_CODE_TEST_GET_POSTAL_CODES_BY_CITY_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "ZIPPOPOTAMUS_ZIP_CODE_TEST_GET_POSTAL_CODES_BY_CITY_ENTID" => idmap,
    "ZIPPOPOTAMUS_ZIP_CODE_TEST_LIVE" => "FALSE",
    "ZIPPOPOTAMUS_ZIP_CODE_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["ZIPPOPOTAMUS_ZIP_CODE_TEST_GET_POSTAL_CODES_BY_CITY_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["ZIPPOPOTAMUS_ZIP_CODE_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
      },
      extra || {},
    ])
    client = ZippopotamusZipCodeSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["ZIPPOPOTAMUS_ZIP_CODE_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["ZIPPOPOTAMUS_ZIP_CODE_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
